### PR TITLE
add edit profile

### PR DESCRIPTION
**Pull request by photonsFardad**
image

image

### CODE REVIEW

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+import { FormikStateSelect, yupStateSchema } from '../utils/States';
+import { FormikTouched, FormikErrors, ErrorMessage, Field, Formik } from 'formik';
+import { Role } from 'packages/sdk/dist/types';
+
+interface EditProfileActionProps {
+  userId: string;
+  onClose: () => void;
+}
+
+const EditProfileActionGetUserQuery = graphql(/* GraphQL */ `
```
[the remark this answers is not in the document]
Comment (marisbest2) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx:32]: Can we use `me` here instead?

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx [hunk 2]
```diff
@@ -2,28 +2,42 @@ import {
...
 
 const profileQuery = graphql(/* GraphQL */ `
   query MeProfileQuery {
     me {
+      id
       name {
+        first
+        last
+        middle
+        title
```
Comment (marisbest2) [inline apps/app/src/views/routes/Settings/components/profile/Profile.tsx:35]: Do we need all these fields at this level?

If I'm reading what you're trying to do correctly, what you want is a fragment, defined down on the form level, and then included up here. And then you'd use `useFragment` in the form. You can roughly copy from `InviteItem` or `UserItem`

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (marisbest2) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx:32]: We should actually be using a fragment here

**Review by marisbest2 — CHANGES_REQUESTED** [review]
We should use a fragment instead of another query. 
And ideally this would work in similar UX to how we have it for Organization

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx (hunk 2 above)
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/Profile.tsx]: done, updated to a common user fragment

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+import { rolesSchema } from '../utils/Roles';
+import { useMutation, useQuery } from '@apollo/client';
+import { useClinicalApiClient } from '../../apollo';
+import { graphql } from 'apps/app/src/gql';
+import { FormikStateSelect, yupStateSchema } from '../utils/States';
+import { FormikTouched, FormikErrors, ErrorMessage, Field, Formik } from 'formik';
+import { Role } from 'packages/sdk/dist/types';
+
+interface EditProfileActionProps {
+  userId: string;
```
[the remark this answers is not in the document]
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx]: removed

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx]: done, updated to a common user fragment

**Comment by thebadprince** [comment]
@photonsFardad Looks good to me. Can we add a confirmation toast that says `Profile updated` after they hit the `Update Profile` button? Similar to how we have it after a user creates a template:

Screenshot 2024-01-09 at 1 33 43 PM

IMO good practice for us UX-wise to always have success/error feedback for the user.

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx [hunk 4]
```diff
@@ -90,10 +112,29 @@ export const Profile = () => {
...
+
+            <Box>
+              <Button size={'sm'} colorScheme={'brand'} leftIcon={<EditIcon />} onClick={onOpen}>
+                Update Profile
+              </Button>
+            </Box>
+          </Flex>
+          <Modal isOpen={isOpen} onClose={onClose} size="xl">
+            <ModalOverlay />
+            {isOpen && (
```
Comment (marisbest2) [inline apps/app/src/views/routes/Settings/components/profile/Profile.tsx:132]: This should only show up if `data` is not null.

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+import { useMutation } from '@apollo/client';
+import { usePhoton } from '@photonhealth/react';
+import { FragmentType, graphql, useFragment } from 'apps/app/src/gql';
+import { FormikStateSelect, yupStateSchema } from '../utils/States';
+import { FormikTouched, FormikErrors, ErrorMessage, Field, Formik } from 'formik';
+import { Role } from 'packages/sdk/dist/types';
+import { userFragment } from '../utils/UserFragment';
+
+interface EditProfileActionProps {
+  user?: FragmentType<typeof userFragment>;
```
Comment (marisbest2) [inline apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx:29]: This should be `user: FragmentType<blah>`. Its always required and non-nullable

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+} from '@chakra-ui/react';
+import * as yup from 'yup';
+import { rolesSchema } from '../utils/Roles';
+import { useMutation } from '@apollo/client';
+import { usePhoton } from '@photonhealth/react';
+import { FragmentType, graphql, useFragment } from 'apps/app/src/gql';
+import { FormikStateSelect, yupStateSchema } from '../utils/States';
+import { FormikTouched, FormikErrors, ErrorMessage, Field, Formik } from 'formik';
+import { Role } from 'packages/sdk/dist/types';
+import { userFragment } from '../utils/UserFragment';
```
[the remark this answers is not in the document]
Comment (marisbest2) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx:26]: It also helps with caching as I _thing_ fragments are cached by name

File: apps/app/src/views/routes/Settings/components/users/UsersList.tsx [hunk 7]
```diff
@@ -35,14 +35,32 @@ const usersQuery = graphql(/* GraphQL */ `
...
       name {
+        first
         full
+        last
+        middle
+        title
       }
       roles {
+        description
+        id
```
Comment (marisbest2) [inline apps/app/src/views/routes/Settings/components/users/UsersList.tsx:61]: I dont think you need any of these fields.

File: apps/app/src/views/routes/Settings/components/utils/UserFragment.tsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import { graphql } from 'apps/app/src/gql';
+export const userFragment = graphql(/* GraphQL */ `
```
Comment (marisbest2) [inline apps/app/src/views/routes/Settings/components/utils/UserFragment.tsx:2]: This should be defined and in the specific files its being used and only specify the specific fields needed. It should also be named something like `<Component>UserFragment` (e.g. `EditRolesActionUserFragment`). Fragments need to be globally uniquely named

**Review by marisbest2 — CHANGES_REQUESTED** [review]
Needs some changes for fragment encapsulation / fragment colocation. (you can read up on Relay-style fragments, though we dont need to go full relay) 

Also, I think it'd be much better UX if its just an inline form, similar to how it works for Organization, rather than a modal.

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx [hunk 9]
```diff
@@ -2,36 +2,56 @@ import {
...
+        first
+        full
+        last
+        middle
+        title
+      }
+      roles {
+        description
+        id
+        name
```
[the remark this answers is not in the document]
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/Profile.tsx:54]: We are they can update them

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx]: done

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx (hunk 4 above)
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/Profile.tsx]: done

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx (hunk 5 above)
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx]: done

File: apps/app/src/views/routes/Settings/components/utils/UserFragment.tsx (hunk 8 above)
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/utils/UserFragment.tsx]: done

Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/utils/UserFragment.tsx]: removed

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/Profile.tsx]: extra fields removed

File: apps/app/src/views/routes/Settings/components/users/UsersList.tsx (hunk 7 above)
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/users/UsersList.tsx]: updated, extra fields removed

File: apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+type ProviderYupType = ProfileYupType['provider'];
+type ProviderFormikTouchedType = FormikTouched<ProviderYupType>;
+type ProviderFormikErrorsType = FormikErrors<ProviderYupType>;
+
+export const EditProfileAction: React.FC<EditProfileActionProps> = ({ user, onClose }) => {
+  const { clinicalClient } = usePhoton();
+  const currentUser = useFragment(userFragment, user);
+  const [updateMyProfile, { error }] = useMutation(UpdateMyProfileMutation, {
+    client: clinicalClient,
+    refetchQueries: ['MeProfileQuery', 'EditProfileActionGetUserQuery']
```
[the remark this answers is not in the document]
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/ProfileEditForm.tsx]: done

**Comment by thebadprince** [comment]
Heads up that the modal title still says: `Assign roles to user`. Should be something like `Edit Profile` 

Screenshot 2024-01-09 at 6 54 00 PM


That said, agree with @marisbest2 here if we can do it:

> Also, I think it'd be much better UX if its just an inline form, similar to how it works for Organization, rather than a modal.

If so, we won't need the title of the modal.

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx [hunk 11]
```diff
@@ -233,6 +235,12 @@ export const Profile = () => {
                   })
             }
           });
+          toast({
+            title: 'Role updated',
```
Comment (thebadprince) [inline apps/app/src/views/routes/Settings/components/profile/Profile.tsx:239]: We'll need this to be `Profile updated` instead of `Role updated`

**Comment by thebadprince** [comment]
Looks good to me aside from the comment above. Thanks for these changes. For some reason, mine is missing Name though:

Screenshot 2024-01-10 at 10 56 32 AM

File: apps/app/src/views/routes/Settings/components/profile/Profile.tsx (hunk 11 above)
Comment (photonsFardad) [reply apps/app/src/views/routes/Settings/components/profile/Profile.tsx]: done

**Review by marisbest2 — APPROVED** [review]

**Pull request merged**